### PR TITLE
Add messaging semantic conventions for settlement spans

### LINKED ISSUE #621: Add conventions for settlement spans for messaging
Based on open-telemetry/oteps#220, conventions for "settle" spans should be added. This involves:

* The "settle" operation.
* Directions of when to create a "settle" spans.
* Defining the relationships of the "settle" span to other spans.

It's not the goal to stabilize those together with messaging semantic conventions. We will work towards a "mixed" stability, where "process" and "settle" spans are experimental, but other parts of the semantic conventions are stable.

The motivation behind is to not delay stabilization efforts with discussions around details, while still having confidence that we can stabilize "process" and "settle" spans without breaking other parts.

### PR DESCRIPTION
**Pull request by pyohannes**
Fixes #621

## Changes

This PR adds conventions around span structure for instrumenting settlement operations in messaging workflows. Those conventions have been agreed on in [OTEP 220](https://github.com/open-telemetry/oteps/blob/main/text/trace/0220-messaging-semantic-conventions-span-structure.md#settlement-spans).

Note that the messaging workgroup agreed to exempt conventions around settlement from initial stability of messaging semantic conventions, those will remain experimental while other parts of semantic conventions will be declared stable.

## Merge requirement checklist

* [x] [CONTRIBUTING.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md) guidelines followed.
* [x] [CHANGELOG.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CHANGELOG.md) updated for non-trivial changes.
* [ ] ~~[schema-next.yaml](https://github.com/open-telemetry/semantic-conventions/blob/main/schema-next.yaml) updated with changes to existing conventions.~~

### CODE REVIEW

File: docs/messaging/messaging-spans.md
Comment (lmolkova) [inline docs/messaging/messaging-spans.md:209]: I think we also need a special-case for settlement span name.

Currently it SHOULD follow `<destination name> <operation name>` pattern, but 

`settle my_queue` is not quite descriptive if more than one settlement kind is supported.
I'd prefer to relax SHOULD and allow systems define how their settlement operations are named.

E.g. RabbitMQ should be able to call them `<destination> ack`  and `<destination> nack` and kafka can use `<destination> commit`. 

I don't see a good reason not to extend it to other spans as well and suggest changing span name part to something like:

```diff
- The values allowed for <operation name> are defined in the section [Operation names](https://github.com/open-telemetry/semantic-conventions/blob/68a87a51e5f4cb3fac9ff029227b73789b462cd7/docs/messaging/messaging-spans.md#operation-names) below. If the format above is used, the operation name MUST match the messaging.operation attribute defined for message consumer spans below.
+ Suggested <operation name> values are defined in the section [Operation names](https://github.com/open-telemetry/semantic-conventions/blob/68a87a51e5f4cb3fac9ff029227b73789b462cd7/docs/messaging/messaging-spans.md#operation-names) below. Messaging systems MAY use their own terminology in span names. For example, RabbitMQ deliver span should be called `<destination> consume` and settle span should be `<destination> ack` or `<destination> nack`
```

We probably need to define span names in extensions for each system.
But this way
- we still have unified `messaging.operation` 
- user-visible name that matches messaging system terminology and API they familiar with 
- we can express settlement type without defining the attribute for it (yet).

Comment (pyohannes) [reply docs/messaging/messaging-spans.md:209]: Discussed in the messaging workgroup: instrumentation SHOULD use `<destination name> <operation name>`. If a messaging system wants to use a different name, they MUST document it in messaging specific conventions.

Comment (pyohannes) [reply docs/messaging/messaging-spans.md:210]: I removed this sentence:
> If the format above is used, the operation name MUST match the `messaging.operation` attribute defined for message consumer spans below.

And I added this one:
> Messaging system specific adaptions to span naming MUST be documented in [semantic conventions for specific messaging technologies](#semantic-conventions-for-specific-messaging-technologies).

Please have another look @open-telemetry/semconv-messaging-approvers

File: docs/messaging/messaging-spans.md
Comment (lmolkova) [inline docs/messaging/messaging-spans.md:277]: not blocking: why MAY and not SHOULD? (which implies 'if possible')

**Review by lmolkova — APPROVED** [review]
LGTM!

File: docs/messaging/messaging-spans.md
Comment (pyohannes) [reply docs/messaging/messaging-spans.md:277]: I took that over from the OTEP.

The MAY gives people more possibilities to model their scenarios. For example [this proposal for GCP Pub/Sub](https://github.com/open-telemetry/semantic-conventions/blob/fb3723986ec78e65e1d6a00eec483ed0db0a9f0f/docs/messaging/gcp-pubsub.md#unary-pull-example) (which still might need some discussion) links settlement-like operations (acks) to the "receive" span. That might be more intuitive in some scenarios?

**Comment by pyohannes** [comment]
@open-telemetry/semconv-messaging-approvers, please have a look.

**Review by joaopgrassi — APPROVED** [review]

File: docs/messaging/messaging-spans.md
Comment (lmolkova) [reply docs/messaging/messaging-spans.md:277]: If I recall we discussed the proposal with @alevenberg and came to the conclusion that link to receive span is MAY, but link to the message is SHOULD. Similar to how we do deliver ---link--> message, but also deliver MAY be a child of message.

This way we have certainty on the relationship - settle is linked to the message (when possible) and flexibility - there could be other links if that's beneficial. 

The assumption is that by default tracing message e2e is important and knowing when/how message was settled is a higher priority than receive/rpc call that received it.

If settle is linked to to the message, it puts it 1 link away. Linking it to receive puts settle 2 links away from the message complicating the queries and joins, and requiring even more luck with sampling (to collect everything).

Comment (pyohannes) [reply docs/messaging/messaging-spans.md:277]: As said above, I kept the MAY because it is in the OTEP, it supports @alevenberg's proposal above, and it maybe also leaves some more room for experimentation during prototyping.

However, I'll bring it up in the next meeting, I'd be fine making it a SHOULD if there's consensus around that. Also, this is something that will be exempt from initial stability, so we can still revisit this after stabilizing conventions.

Comment (lmolkova) [reply docs/messaging/messaging-spans.md:277]: sure, let's not block on this

Comment (alevenberg) [reply docs/messaging/messaging-spans.md:277]: Not blocking: but the reasoning for linking to the receive span instead of the create span is if there are multiple subscription resources.

For example, if a message is published to topic-a and then received on sub-a and sub-b, you would want the settle span to link to the corresponding subscription span directly.

Comment (lmolkova) [reply docs/messaging/messaging-spans.md:277]: I think we would achieve a similar effect by setting subscription name attribute on a settlement span. 

without deliver or process span user experience would not be great either way and with it links wouldn't matter that much. So it's definitely not a hill I want to die on :)

Comment (alevenberg) [reply docs/messaging/messaging-spans.md:277]: Yeah I think the issue was if it is a parent/child relationship between the create and receive and it had multiple subs if we link to the create, it's not as clear what sub it was a part of. 

I think keeping it as MAY is nice to give the user the flexibility depending on their instrumentation.

File: docs/messaging/messaging-spans.md
Comment (hongalex) [inline docs/messaging/messaging-spans.md:198]: Google Pub/Sub plans to differ from this naming slightly. `extend` and `settle` are replaced with `modack` (Modify Ack Deadline) and `ack` (acknowledge).

The concept of settlement is not known within our documentation so we're worried it may confuse users. This will be documented in the section described though.

Comment (joaopgrassi) [reply docs/messaging/messaging-spans.md:198]: @hongalex so are you good with the way it is structured/mentioned now? Or you think we need to change something?

Comment (pyohannes) [reply docs/messaging/messaging-spans.md:198]: @hongalex Yes, that's exactly what this clause is for. You can use different span and operation names, as long as you document it in the system-specific portions of the conventions.

Comment (hongalex) [reply docs/messaging/messaging-spans.md:198]: Yeah I just wanted to provide heads up as to why the naming might look different.

The actual language looks good though!

**Review by alevenb — APPROVED** [review]

**Review by hongalex — APPROVED** [review]

**Comment by joaopgrassi** [comment]
Hi @pyohannes !

We changed how the CHANGELOG.md is managed. Please take a look at https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md#adding-a-changelog-entry to see what needs to be done. Sorry for the disruption.

**Review by jsuereth — APPROVED** [review]

**Pull request merged**
